content/blog/2016/2016-04-07-pipeline-for-runs-on-hardware.adoc

```diff
@@ -0,0 +1,238 @@
+It can be easily done by a _Execute shell_ build steps in free-style projects.
+Such testing scheme is commonly affected by the following issues:
+
+* Nodes with peripherals are being shared across several projects. 
+Jenkins must insure the correctness of access (e.g. by throttling the access). 
+** In a single Freestyle project builds utilize the node for a long period. If you synthesize the item before the run, much of the peripheral utilization file may be wasted.
+** The issue can be solved by one of concurrency management plugins:
+link:https://wiki.jenkins-ci.org/display/JENKINS/Throttle+Concurrent+Builds+Plugin[Throttle Concurrent Builds], link:https://wiki.jenkins-ci.org/display/JENKINS/Lockable+Resources+Plugin[Lockable Resources]
+ or 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Exclusion-Plugin[Exclusions].
+* Test parallelization on multiple nodes requires using of multiple projects or 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Matrix+Project+Plugin[Matrix configurations], so it causes job chaining again.
+** These build chains can be created via 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Parameterized+Trigger+Plugin[Parameterized Trigger] and 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Copy+Artifact+Plugin[Copy Artifacts], but it complicates job management and build history investigation.
+* Hardware infrastructure is usually flaky. 
+If it fails during the build due to any reason, it's hard to diagnose the issue and re-run the project if the issue comes from hardware.
+** link:https://wiki.jenkins-ci.org/display/JENKINS/Build+Failure+Analyzer[Build Failure Analyzer] allows to identify the root cause of a build failure (e.g. by build log parsing).
+** link:https://wiki.jenkins-ci.org/display/JENKINS/Conditional+BuildStep+Plugin[Conditional Build Step] and 
+link:https://wiki.jenkins-ci.org/display/JENKINS/Flexible+Publish+Plugin[Flexible Publish] plugins allow altering the build flow according to the analysis results.
+** Combination of the plugins above is possible, but it makes job configurations extremely large.
+* Tests on hardware peripherals may take much time. 
+If an infrastructure fails, we may have to restart the run from scratch. 
+So the builds should be robust against infra issues including network failures and Jenkins master restarts.
+
+Usage of 
+link:https://jenkins.io/doc/pipeline/[Jenkins Pipeline]  can effectively address the use-cases above.
+In this blogpost I want to address several cases.
+
+## Pipeline-as-Code for test runs on hardware
+
+Pipeline-as-Code is an emerging technology for representing job configurations as code.
+In Jenkins there are two dominant implementations - Pipeline and JobDSL plugins.
+JobDSL Plugin internally generates common freestyle jobs according to the script, so it's functionality is similar to the description above.
+In this post I will concentrate on the Pipeline plugin.
```

Comment:
concentrate -> focus (in French, I would also write concentrate, but not in English!)